rosbag2_storage/src/rosbag2_storage/qos.cpp

```diff
@@ -149,57 +149,58 @@ bool convert<rmw_time_t>::decode(const Node & node, rmw_time_t & time)
   return true;
 }
 
-Node convert<rosbag2_storage::Rosbag2QoS>::encode(
-  const rosbag2_storage::Rosbag2QoS & qos)
+Node convert<rclcpp::QoS>::encode(const rclcpp::QoS & qos, int version)
 {
   const auto & p = qos.get_rmw_qos_profile();
   Node node;
-  node["history"] = convert<rmw_qos_history_policy_t>::encode(p.history, 9);
+  node["history"] = convert<rmw_qos_history_policy_t>::encode(p.history, version);
   node["depth"] = p.depth;
-  node["reliability"] = convert<rmw_qos_reliability_policy_t>::encode(p.reliability, 9);
-  node["durability"] = convert<rmw_qos_durability_policy_t>::encode(p.durability, 9);
+  node["reliability"] = convert<rmw_qos_reliability_policy_t>::encode(p.reliability, version);
+  node["durability"] = convert<rmw_qos_durability_policy_t>::encode(p.durability, version);
   node["deadline"] = p.deadline;
   node["lifespan"] = p.lifespan;
-  node["liveliness"] = convert<rmw_qos_liveliness_policy_t>::encode(p.liveliness, 9);
+  node["liveliness"] = convert<rmw_qos_liveliness_policy_t>::encode(p.liveliness, version);
   node["liveliness_lease_duration"] = p.liveliness_lease_duration;
   node["avoid_ros_namespace_conventions"] = p.avoid_ros_namespace_conventions;
   return node;
 }
 
-bool convert<rosbag2_storage::Rosbag2QoS>::decode(
-  const Node & node, rosbag2_storage::Rosbag2QoS & qos, int version)
+bool convert<rclcpp::QoS>::decode(const Node & node, rclcpp::QoS & qos, int version)
 {
   rmw_qos_history_policy_t history;
   rmw_qos_reliability_policy_t reliability;
   rmw_qos_durability_policy_t durability;
   rmw_qos_liveliness_policy_t liveliness;
 
-  int temp;
-  if (convert<int>::decode(node["history"], temp)) {
-    history = static_cast<rmw_qos_history_policy_t>(temp);
+  // Try to auto-detect qos serialization format
+  int history_int = -1;
+  if (convert<int>::decode(node["history"], history_int) &&
+    history_int >= RMW_QOS_POLICY_HISTORY_SYSTEM_DEFAULT &&
+    history_int <= RMW_QOS_POLICY_HISTORY_UNKNOWN)
+  {
+    history = static_cast<rmw_qos_history_policy_t>(history_int);
+    version = 8;
   } else {
     history = node["history"].as<rmw_qos_history_policy_t>();
+    version = 9;
   }
 
   if (version <= 8) {
+    history = static_cast<rmw_qos_history_policy_t>(node["history"].as<int>());
     reliability = static_cast<rmw_qos_reliability_policy_t>(node["reliability"].as<int>());
```

Comment:
Shouldn't we use the same auto-detect strategy for reliability, durability and liveliness?

Reply:
I am using history to automatically detect version number
```
  // Try to auto-detect qos serialization format
  int history_int = -1;
  if (convert<int>::decode(node["history"], history_int) &&
    history_int >= RMW_QOS_POLICY_HISTORY_SYSTEM_DEFAULT &&
    history_int <= RMW_QOS_POLICY_HISTORY_UNKNOWN)
  {
    history = static_cast<rmw_qos_history_policy_t>(history_int);
    version = 8;
  } else {
    history = node["history"].as<rmw_qos_history_policy_t>();
    version = 9;
  }
```

However, the line after that with the history assignment could be removed. I forgot to remove it
```
if (version <= 8) {
    history = static_cast<rmw_qos_history_policy_t>(node["history"].as<int>());
```

Reply:
No problem, will remove it then :)